guides/api-style.rst

```diff
@@ -0,0 +1,118 @@
+  hypothesis.extra module.
+* Complexity should not be pushed onto the user. An easy to use API is more
+  important than a simple implementation.
+
+~~~~~~~~~~~~~~~~~~~~~~~~~
+Guidelines for strategies
+~~~~~~~~~~~~~~~~~~~~~~~~~
+
+* A strategy function should be somewhere between a recipe for how to build a
+  value and a range of valid values.
+* It should not include distribution hints. The arguments should only specify
+  how to produce a valid value, not statistical properties of values.
+* Strategies should try to paper over non-uniformity in the underlying types
+  as much as possible (e.g. ``hypothesis.extra.numpy`` has a number of
+  workarounds for numpy's odd behaviour around object arrays).
+
+~~~~~~~~~~~~~~~~~
+Argument handling
+~~~~~~~~~~~~~~~~~
+
+We have a reasonably distinctive style when it comes to handling arguments:
+
+* Arguments must be validated to the greatest extent possible. Hypothesis
+  should reject bad arguments with an InvalidArgument error, not fail with an
+  internal exception.
+* We make extensive use of default arguments. If an argument could reasonably
+  have a default, it should.
+* Exception to the above: Strategies for collection types should *not* have a
+  default argument for element strategies.
+* Interacting arguments (e.g. arguments that must be in a particular order, or
+  where at most one is valid, or where one argument restricts the valid range
+  of the other) are fine, but when this happens the behaviour of defaults
+  should automatically be adjusted. e.g. if the normal default of an argument
+  would become invalid, the function should still do the right thing if that
+  default is used.
+* Where the actual default used depends on other arguments, the default parameter
+  should be None.
+* It's worth thinking about the order of arguments: the first one or two
+  arguments are likely to be passed positionally, so try to put values there
+  where this is useful and not too confusing.
+* Arguments should not be "a value or a strategy for generating that value"
+  unless there is some highly compelling reason for that. Hypothesis has good
+  composition mechanisms and people should use them instead of adding them on
+  an ad hoc basis to individual API functions.
+
+~~~~~~~~~~~~~~
+Function Names
+~~~~~~~~~~~~~~
+
+We don't have any real consistency here. The rough approach we follow is:
+
+* Names are `snake_case` as is standard in Python.
+* Strategies for a particular type are typically named as a plural name for
+  that type. Where that type has some truncated form (e.g. int, str) we use a
+  longer form name.
+* Other strategies have no particular common naming convention.
+
+~~~~~~~~~~~~~~
+Argument Names
+~~~~~~~~~~~~~~
+
+We should try to use the same argument names and orders across different
+strategies wherever possible. In particular:
+
+* For collection types, the element strategy (or strategies) should always be
+  the first argument. It does not have a specific common name.
```

Comment:
Agreed; but mostly I meant that `elements` is in fact the common name for this argument to collection strategies.